packages/xstate-react/src/useMachine.ts

```diff
@@ -91,28 +65,67 @@ export function useMachine<TMachine extends AnyStateMachine>(
   getMachine: MaybeLazy<TMachine>,
   ...[options = {}]: RestParams<TMachine>
 ): UseMachineReturn<TMachine> {
-  const listener = useCallback((nextState: StateFrom<TMachine>) => {
-    // Only change the current state if:
-    // - the incoming state is the "live" initial state (since it might have new actors)
-    // - OR the incoming state actually changed.
-    //
-    // The "live" initial state will have .changed === undefined.
-    const initialStateChanged =
-      nextState.changed === undefined && Object.keys(nextState.children).length;
+  // using `useIdleInterpreter` allows us to subscribe to the service *before* we start it
+  // so we don't miss any notifications
+  const service = useIdleInterpreter(getMachine, options as any);
 
-    if (nextState.changed || initialStateChanged) {
-      setState(nextState);
+  const getSnapshot = useCallback(() => {
+    if (service.status === InterpreterStatus.NotStarted) {
+      return (options.state
+        ? State.create(options.state)
+        : service.machine.initialState) as State<any, any, any, any, any>;
     }
-  }, []);
 
-  const service = useInterpret(getMachine as any, options as any, listener);
+    return service.state;
+  }, [service]);
 
-  const [state, setState] = useState(() => {
-    const { initialState } = service.machine;
-    return (options.state
-      ? State.create(options.state as any)
-      : initialState) as StateFrom<TMachine>;
-  });
+  const isEqual = useCallback(
+    (prevState, nextState) => {
+      if (service.status === InterpreterStatus.NotStarted) {
+        return true;
+      }
+
+      // Only change the current state if:
+      // - the incoming state is the "live" initial state (since it might have new actors)
+      // - OR the incoming state actually changed.
+      //
+      // The "live" initial state will have .changed === undefined.
+      const initialStateChanged =
+        nextState.changed === undefined &&
+        (Object.keys(nextState.children).length > 0 ||
+          typeof prevState.changed === 'boolean');
```

Comment:
this additional condition here allows us to "restart" to the initial state after Fast Refresh, whereas previously we were displaying the outdated state that was no longer true for the current, restarted, service